jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,6 +192,433 @@ default void clear()
     }
 
+    /** Append byte array to the buffer, potentially limited by capacity.
+     * @param bytes the byte array to append
+     * @param offset the offset into the array
+     * @param length the number of bytes to try to append
+     * @return the number of bytes actually appended.
+     */
+    default int append(byte[] bytes, int offset, int length)
+    {
+        ByteBuffer to = getByteBuffer();
+
+        int pos = BufferUtil.flipToFill(to);
+        try
+        {
+            length = Math.min(length, to.remaining());
+            to.put(bytes, offset, length);
+        }
+        finally
+        {
+            BufferUtil.flipToFlush(to, pos);
+        }
+        return length;
+    }
+
+    default boolean append(ByteBuffer bytes)
+    {
+        BufferUtil.append(getByteBuffer(), bytes);
+        return !bytes.hasRemaining();
+    }
+
+    default boolean append(RetainableByteBuffer bytes)
+    {
+        return bytes.remaining() == 0 || append(bytes.getByteBuffer());
+    }
+
+    default void putTo(ByteBuffer toInfillMode)
+    {
+        toInfillMode.put(getByteBuffer());
+    }
+
+    default void writeTo(OutputStream out) throws IOException
+    {
+        BufferUtil.writeTo(getByteBuffer(), out);
+    }
+
+    default void writeTo(Content.Sink sink, boolean last, Callback callback)
+    {
+        sink.write(last, getByteBuffer(), callback);
+    }
+
+    /**
+     * An aggregating {@link RetainableByteBuffer} that may grow when content is appended to it.
+     */
+    class Aggregator implements RetainableByteBuffer
+    {
+        private final ByteBufferPool _pool;
+        private final boolean _direct;
+        private final int _growBy;
+        private final int _maxCapacity;
+        private RetainableByteBuffer _buffer;
+
+        /**
+         * Construct an aggregating {@link RetainableByteBuffer} that may grow when content is appended to it.
+         * {@link RetainableByteBuffer}s with zero-copy if the {@link #append(RetainableByteBuffer)} API is used
+         * @param pool The pool from which to allocate buffers
+         * @param direct true if direct buffers should be used
+         * @param maxCapacity The maximum requested length of the accumulated buffers or -1 for no limit.
+         *                    Note that the pool may provide a buffer that exceeds this capacity.
+         */
+        public Aggregator(ByteBufferPool pool, boolean direct, int maxCapacity)
+        {
+            this(pool, direct, -1, maxCapacity);
+        }
+
+        /**
+         * Construct an aggregating {@link RetainableByteBuffer} that may grow when content is appended to it.
+         * {@link RetainableByteBuffer}s with zero-copy if the {@link #append(RetainableByteBuffer)} API is used
+         * @param pool The pool from which to allocate buffers
+         * @param direct true if direct buffers should be used
+         * @param growBy the size to grow the buffer by or &lt;= 0 for a heuristic
```

Comment:
I have a feeling that `growBy` and `maxCapacity` should not have heuristics nor defaults, as I think those would need to be specific depending on what uses the aggregator.

Reply:
I agree the heuristic is not the best.   But it is just as hard for the caller to know the best value.  It really should be the buffer pools bucket increment.    Let me see if I can at least come up with a better heuristic.